Homely

kolaf opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add support for the Norwegian company Homely. They are a sort of a DIY alarm company that provides you with a Zigbee gateway they controll, together with any number of devices you care to install in the house. They provide a simple rest and web socket API to pull and push data from the various sensors. There is currently no API for controlling devices (e.g. arming the alarm). 

This is an initial PR that supports populating the devices and entities, and polling for an updated state of these every five minutes. This will eventually be changed to use the web socket interface and do away with the polling. There is also limited support for devices, this is limited by the devices I actually have in my own house.

Is this the correct place for this, or should it perhaps be added to HACS instead?

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]: https://github.com/home-assistant/home-assistant.io/pull/25502

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

kolaf committed:
Initial commit with configuration flow

kolaf committed:
Created some devices and entities

kolaf committed:
Most binary sensors are working

kolaf committed:
Fixes battery low error. Adds temperature, signal strength and battery voltage sensors.

kolaf committed:
Removes commented code

kolaf committed:
Code cleanup

kolaf committed:
Merge branch 'home-assistant:dev' into homely

kolaf committed:
Bumps homely to 0.0.3. Fixes config flow test.

kolaf committed:
Merge branch 'homely' of https://github.com/kolaf/core into homely

kolaf committed:
Bumps homely to 0.0.4. Fixes strings in flow.

kolaf committed:
Adds Homely to coverage ignore file.

kolaf committed:
Merge branch 'dev' into homely

Kane610 reviewed (approved):
I think this looks really good, couldn't see any issues myself.

kolaf commented on homeassistant/components/homely/binary_sensor.py:
```diff
@@ -0,0 +1,113 @@
+class HomelyBinarySensorEntity(CoordinatorEntity, BinarySensorEntity):
+    """Abstract binary sensor class."""
+
+    _attr_name = "Basic"
```
Do you mean the name attribute or the full class? I guess you mean the name attribute, and I agree that this can probably be removed.

kolaf replied:
Removed from both sensor and binary_sensor

kolaf committed:
Removes commented code

Co-authored-by: Robert Svensson <Kane610@users.noreply.github.com>

kolaf committed:
Removes unused _attr_name

kolaf committed:
Merge branch 'homely' of https://github.com/kolaf/core into homely

kolaf committed:
Fixes docstring

bdraco commented on homeassistant/components/homely/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/homely",
+  "requirements": ["homelypy==0.0.7"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
```
```suggestion
```

bdraco commented on homeassistant/components/homely/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  "zeroconf": [],
+  "homekit": {},
+  "integration_type": "hub",
+  "dependencies": [],
```
```suggestion
```

bdraco commented on homeassistant/components/homely/sensor.py:
```diff
@@ -0,0 +1,105 @@
+    coordinator = homely_home.coordinator
+    await coordinator.async_config_entry_first_refresh()
+
+    entities: list[SensorEntity] = []
+    for homely_device in homely_home.devices.values():
+        entities.append(TemperatureEntity(coordinator, homely_device))
+        entities.append(SignalStrengthEntity(coordinator, homely_device))
+        entities.append(BatteryVoltageEntity(coordinator, homely_device))
+    async_add_entities(entities)
```
This could be a comprehension

kolaf replied:
Change to list comprehension

bdraco commented on tests/components/homely/test_config_flow.py:
```diff
@@ -0,0 +1,97 @@
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "cannot_connect"}
```
Please finish config flow tests all the way to an ABORT or CREATE_ENTRY state so we know it can recover

kolaf replied:
The test now runs the successful config flow after the initial failure.

bdraco commented on tests/components/homely/test_config_flow.py:
```diff
@@ -0,0 +1,97 @@
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "invalid_auth"}
```
Please finish config flow tests all the way to an ABORT or CREATE_ENTRY state so we know it can recover

kolaf replied:
The test now runs the successful config flow after the initial failure.

bdraco commented on homeassistant/components/homely/homely_device.py:
```diff
@@ -0,0 +1,132 @@
+    async def setup(self) -> None:
+        """Perform initial setup."""
+        try:
+            self.homely = Homely(self.username, self.password)
```
Can this raise an exception?  If not please move it outside the try block

kolaf replied:
No, so moved outside the try block.

bdraco commented on homeassistant/components/homely/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+            )
+        schema = vol.Schema(
+            {
+                vol.Required("location",): vol.In(
```
```suggestion
                vol.Required("location"): vol.In(
```

bdraco commented on homeassistant/components/homely/__init__.py:
```diff
@@ -0,0 +1,37 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.BINARY_SENSOR, Platform.SENSOR]
```
Usually we only accept a single platform for new integration merges
https://developers.home-assistant.io/docs/creating_component_code_review?_highlight=single&_highlight=platform#5-make-your-pull-request-as-small-as-possible

kolaf replied:
I am not sure what to do about this. Is it okay as it is, or should I remove one of the platforms and put in a second pull request afterwards?

frenck replied:
Correct, only a single platform in the initial PR is allowed

kolaf replied:
I have removed the binary sensor platform.

kolaf committed:
Update homeassistant/components/homely/manifest.json

Co-authored-by: J. Nick Koston <nick@koston.org>

kolaf committed:
Update homeassistant/components/homely/manifest.json

Co-authored-by: J. Nick Koston <nick@koston.org>

kolaf committed:
Update homeassistant/components/homely/config_flow.py

Co-authored-by: J. Nick Koston <nick@koston.org>

kolaf committed:
Runs config flow tests to completion. Minor code review refactoring

frenck commented:
The linked documentation PR is incorrect and has to be re-created. Adding docs-missing flag for now.

frenck commented on homeassistant/components/homely/__init__.py:
```diff
@@ -0,0 +1,37 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up homely from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+    homely_home = HomelyHome(hass, entry)
+    await homely_home.setup()
+    hass.data[DOMAIN][entry.entry_id] = homely_home
```
```suggestion
    homely_home = HomelyHome(hass, entry)
    await homely_home.setup()
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = homely_home
```

frenck commented on homeassistant/components/homely/binary_sensor.py:
```diff
@@ -0,0 +1,112 @@
+        self.async_write_ha_state()
+
+
+class WindowSensorEntity(HomelyBinarySensorEntity):
+    """Represent a window sensor."""
+
+    _attr_name = "Window"
+    _attr_device_class = BinarySensorDeviceClass.WINDOW
+
+
+class MotionSensorEntity(HomelyBinarySensorEntity):
+    """Represent a motion sensor."""
+
+    _attr_name = "Motion"
+    _attr_device_class = BinarySensorDeviceClass.MOTION
+
+
+class SmokeAlarmEntity(HomelyBinarySensorEntity):
+    """Represent a smoke alarm."""
+
+    _attr_name = "Smoke"
+    _attr_device_class = BinarySensorDeviceClass.SMOKE
+
+    def get_is_on_from_device(self, device: Device) -> bool:
+        """Get the current state of the sensor."""
+        return device.alarm.fire
+
+
+class BatteryLowEntity(HomelyBinarySensorEntity):
+    """Represent battery low state."""
+
+    _attr_name = "Battery low"
+    _attr_device_class = BinarySensorDeviceClass.BATTERY
+
+    def get_is_on_from_device(self, device: Device) -> bool:
+        """Get the current state of the sensor."""
+        return device.battery.low
```
Instead of creating all these classes, please use `BinarySensorEntityDescription`s instead.

kolaf replied:
I see how this may solve the device class issue, but how do I handle the different ways of fetching the entity state if not through subclasses?

frenck replied:
You could add a lamba to the entity description to handle specific cases. (search for `value_fn` in our codebase for examples).

kolaf replied:
Moved to device specific logic to entity descriptions

frenck reviewed (changes requested):
Some initial comments, will do another pass once these have been addressed.

kolaf committed:
Update homeassistant/components/homely/__init__.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

kolaf committed:
Moves coordinator into HomelyHome class

kolaf commented on homeassistant/components/homely/homely_device.py:
```diff
@@ -0,0 +1,132 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class HomelyHome:
```
Moved to coordinator.py and redefined as a coordinator. The only place this is instantiated is in __init__.py.

kolaf commented on homeassistant/components/homely/homely_device.py:
```diff
@@ -0,0 +1,132 @@
+    async def _async_update_data(self) -> None:
+        self.location = await self.hass.async_add_executor_job(
+            self.homely.get_location, self.location.location_id
+        )
```
Merged with HomelyHome and moved to correct file.

kolaf committed:
Removes unnecessary first entry refresh

kolaf commented on homeassistant/components/homely/binary_sensor.py:
```diff
@@ -0,0 +1,112 @@
+    homely_home: HomelyHome = hass.data[DOMAIN][entry.entry_id]
+
+    coordinator = homely_home.coordinator
+    await coordinator.async_config_entry_first_refresh()
```
Removed from the sensors and placed in __init__.py.

kolaf commented on homeassistant/components/homely/homely_device.py:
```diff
@@ -0,0 +1,132 @@
+        return PollingDataCoordinator(self.hass, self.homely, self.location)
+
+
+class HomelyDevice:
```
No, this is a device. I have cleaned up the class to make it clearer that it basically supplies the device_info that is used by the various entities.

kolaf committed:
Switch to using entity descriptions

kolaf committed:
Removes binary sensor platform

kolaf commented:
Quick question. When resolving an issue pointed out in a comment, do I mark the comment as resolved or wait for the reviewer to do it?

kolaf committed:
Adds support for HAN power metre device. Removes binary_sensor from .coveragerc

kolaf committed:
Merge branch 'home-assistant:dev' into homely

kolaf committed:
Fixes error in dependency and bumps version

kolaf committed:
Merge branch 'homely' of https://github.com/kolaf/core into homely

kolaf commented:
I got some notifications that some test runs had failed. This was caused by an error in my API package and I have pushed a commit that should fix this. How do I trigger a new test run?

kolaf committed:
Bumps homelypy version for better error handling

kolaf commented:
Is there anything more required from the for this to move on? I'm not very experienced with public pull requests like this, so I'm not entirely sure what the correct flow is. Thank you very much for your time and effort in helping me.

kolaf committed:
Fixes metering sensor scaling.

kolaf committed:
Merge branch 'dev' into homely

kolaf commented:
@frenck I don't want to be pushy, but I think I addressed all of your comments some time ago. I would appreciate it greatly if you could take another look and let me know if there is anything still missing. Thanks for your help.

I have a separate branch that is almost complete that adds the remaining platforms and also add support for the streaming API which is much better for alarm and door sensors than the polling approach used in this pull request. I hope this PR can be completed soon so that we can move forward with the next branch which is required to have a fully featured Homely integration.

kolaf committed:
Merge branch 'dev' into homely

kolaf commented:
@frenck I am still waiting for you to review my responses to your requested changes. Could you please help me out with this.

MartinHjelmare commented on .coveragerc:
```diff
@@ -466,6 +466,12 @@ omit =
     homeassistant/components/home_connect/switch.py
     homeassistant/components/home_plus_control/api.py
     homeassistant/components/home_plus_control/switch.py
+<<<<<<< homely
```
Please solve the remaining merge conflict marker.

MartinHjelmare commented on homeassistant/components/homely/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+        if user_input is not None:
+            location_id: str = user_input["location"]
+            location = next(
+                filter(
```
Please use a generator expression instead of filter + lambda.

MartinHjelmare commented on homeassistant/components/homely/coordinator.py:
```diff
@@ -0,0 +1,57 @@
+            raise ConfigEntryNotReady(f"Unable to connect to Homely: {ex}") from ex
+        await self.update_devices()
+
+    async def update_devices(self) -> None:
```
This doesn't have to be a coroutine function since there's no await inside.

MartinHjelmare commented on homeassistant/components/homely/coordinator.py:
```diff
@@ -0,0 +1,57 @@
+        """Perform initial setup."""
+        self.homely = Homely(self.username, self.password)
+        try:
+            self.location = await self.hass.async_add_executor_job(
```
We can remove the `setup` function as we get the location in `_async_update_data` too.

MartinHjelmare commented on homeassistant/components/homely/coordinator.py:
```diff
@@ -0,0 +1,57 @@
+            )
+        except (ConnectionFailedException, ConnectTimeout, HTTPError) as ex:
+            raise ConfigEntryNotReady(f"Unable to connect to Homely: {ex}") from ex
+        await self.update_devices()
```
Update the devices if the stored devices dict is empty in `_async_update_data` instead.

Are we sure we don't need to update devices on every update? How is the device state updated?

MartinHjelmare commented on homeassistant/components/homely/coordinator.py:
```diff
@@ -0,0 +1,57 @@
+
+    async def setup(self) -> None:
+        """Perform initial setup."""
+        self.homely = Homely(self.username, self.password)
```
We can create the client when creating the coordinator in the init method.

MartinHjelmare commented on homeassistant/components/homely/homely_device.py:
```diff
@@ -0,0 +1,53 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class HomelyDevice:
```
This class doesn't really seem needed. It's just proxying the same attributes as the library device exposes.

MartinHjelmare commented on homeassistant/components/homely/homely_device.py:
```diff
@@ -0,0 +1,53 @@
+        return self._device
+
+    @property
+    def device_info(self) -> DeviceInfo:
```
The device info can be defined in a common entity base class instead.
